### ISSUE TITLE
Provide alternative nvs_init that will take const struct *device instead of device name

### ISSUE THREAD

**Issue by de-nordic** [zephyrproject-rtos/zephyr#33627]
**Is your enhancement proposal related to a problem? Please describe.**
There is ongoing change to flash API that introduces `fa_dev` field, of type `const struct device *`, to the `flash_area` structure (
https://github.com/zephyrproject-rtos/zephyr/pull/33346). That pointer will be used internally by the flash API to obtain access to device tied to the flash area, instead of doing so at run-time with call to `device_get_binding`.
The `nvs_init` should be also switched to using the pointer instead of calling `device_get_binding` as it does so now, with second parameter that is actual device name.
The `fa_dev`, for default flash map, is resolved in compile-time which is faster and does not require iterating the list of devices to find the one by name.
Additionally `fa_dev_name` field of `flash_area` becomes optional field, controlled by Kconfig option, but in case when `nvs_init` is used in current form it will be required.

**Describe the solution you'd like**
Provide alternative `nvs_init` function that will take `const struct device *` type second parameter and will use provided device object instead of trying to figure it out itself by the name of flash device.
The additional `nvs_init` is desired, instead of replacing the second parameter of the original `nvs_init`, due to possibility of the function being used in user code that is not within Zephyr source tree.
The original function `nvs_init` may be optional, depending on existence of `flash_area.fa_dev_name` and deprecated in the future.

**Describe alternatives you've considered**
Replacing the `nvs_init` second parameter with `const struct device *` type, changing the function signature,  and possibly breaking user code.

**Comment by nvlsianpu** [comment]
I don't have strong preference. We can introduce nvs_init_with_dev() or something like that.

**Comment by Laczen** [comment]
Maybe there is no need to make a new function. If nvs_init() is called with `fs.flash_device` already defined the call to `get_device_binding` could be skipped.

**Comment by nvlsianpu** [comment]
@Laczen I think this would be better to have API extended or modified.
What you have suggested will works, but will be looking as a hack:

``` C
fs.flash_device = flash_area_get_device(&fa);
rc = nvs_init(&fs, NULL);
```

I would either change nvs_init()
``` C
int nvs_init(struct nvs_fs *fs)
```
or add new API: 
``` C
int nvs_init_pure(struct nvs_fs *fs)
```

Are you are reluctant to API update?

**Comment by Laczen** [comment]
@nvlsianpu, I am not reluctant to API update but would make the transition as effortless as possible:
a. We keep `nvs_init()` as is and give a warning when it is not called with *dev_name=NULL a warning is given that it should be updated.
b. After some time a switch is made to define as `nvs_init(struct nvs_fs *fs)`

Instead of using the `flash_area` api I would just use `fs.flash_device = DEVICE_DT_GET(...)`

**Comment by martinjaeger** [comment]
Looks like this has been implemented in #43260 by @gmarull. Please re-open if you disagree.
